---
title: "Running fastMNN on Seurat Objects"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: 
  github_document:
    html_preview: true
    toc: true
    toc_depth: 3
    fig_width: 16
  html_document:
    df_print: kable
    theme: united
    fig_height: 5
    fig_width: 16
    out_height: 4
---

This vigettte demonstrates how to run fastMNN on Seurat objects. Parameters and commands are based off of the [fastMNN help page](https://rdrr.io/github/LTLA/batchelor/man/fastMNN.html). If you use fastMNN, please cite:

> *Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors*
>
> Laleh Haghverdi, Aaron T L Lun, Michael D Morgan & John C Marioni
>
> Nature Biotechnology, 2018
>
> doi: [10.1038/nbt.4091](https://doi.org/10.1038/nbt.4091)
>
> Bioconductor: https://bioconductor.org/packages/release/bioc/html/batchelor.html

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE
)
options(timeout = 1000)
```

Prerequisites to install: 

* [Seurat](https://satijalab.org/seurat/install)
* [SeuratWrappers](https://github.com/satijalab/seurat-wrappers)
* [SeuratData](https://github.com/satijalab/seurat-data)

```{r packages}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
```

## {.tabset .tabset-pills}

### Systematic comparative analysis of human PBMC

To learn more about this dataset, type `?pbmcsca` 

```{r pbmcsca, results='hide'}
InstallData("pbmcsca")
data("pbmcsca")
pbmcsca <- NormalizeData(pbmcsca)
pbmcsca <- FindVariableFeatures(pbmcsca)
pbmcsca <- RunFastMNN(object.list = SplitObject(pbmcsca, split.by = 'Method'))
pbmcsca <- RunUMAP(pbmcsca, reduction = 'mnn', dims = 1:30)
pbmcsca <- FindNeighbors(pbmcsca, reduction = 'mnn', dims = 1:30)
pbmcsca <- FindClusters(pbmcsca)
DimPlot(pbmcsca, group.by = c('Method', 'ident',"CellType"), ncol = 3)
```

```{r cleanup_pbmcsca, include=FALSE}
rm(pbmcsca)
gc(verbose = FALSE)
```

### Interferon-stimulated and control PBMC

To learn more about this dataset, type `?ifnb` 

```{r ifnb_stim, results='hide'}
InstallData("ifnb")
data("ifnb")
ifnb <- NormalizeData(ifnb)
ifnb <- FindVariableFeatures(ifnb)
ifnb <- RunFastMNN(object.list = SplitObject(ifnb, split.by = 'stim'))
ifnb <- RunUMAP(ifnb, reduction = 'mnn', dims = 1:30)
ifnb <- FindNeighbors(ifnb, reduction = 'mnn', dims = 1:30)
ifnb <- FindClusters(ifnb)
DimPlot(ifnb, group.by = c('stim', 'ident', 'seurat_annotations'), ncol = 3)
```

```{r cleanup_ifnb, include=FALSE}
rm(ifnb)
gc(verbose = FALSE)
```

### Eight human pancreatic islet datasets

To learn more about this dataset, type `?panc8` 

```{r pancreas, results='hide'}
InstallData("panc8")
data("panc8")
panc8 <- NormalizeData(panc8)
panc8 <- FindVariableFeatures(panc8)
panc8 <- RunFastMNN(object.list = SplitObject(panc8, split.by = 'replicate')[c("celseq", "celseq2", "fluidigmc1", "smartseq2")])
panc8 <- RunUMAP(panc8, reduction = 'mnn', dims = 1:30)
panc8 <- FindNeighbors(panc8, reduction = 'mnn', dims = 1:30)
panc8 <- FindClusters(panc8)
DimPlot(panc8, group.by = c('replicate', 'ident', 'celltype'), ncol = 3) 
```
