<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Analyzing Seurat data using fastTopics</title>

<script src="fasttopics_files/header-attrs-2.14/header-attrs.js"></script>
<script src="fasttopics_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="fasttopics_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="fasttopics_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="fasttopics_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="fasttopics_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="fasttopics_files/navigation-1.1/tabsets.js"></script>
<link href="fasttopics_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="fasttopics_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Analyzing Seurat data using fastTopics</h1>
<h4 class="date">Compiled: March 17, 2023</h4>

</div>


<p>&lt;– Run this to generate the HTML and Markdown outputs:
library(rmarkdown) render(“fasttopics.Rmd”)
render(“fasttopics.Rmd”,output_format = md_document()) –&gt;</p>
<p>This vignette illustrates the use of the <a
href="https://github.com/stephenslab/fastTopics">fastTopics</a> Seurat
wrapper to analyze a Seurat data set. This vignette is intended only to
introduce the basic fastTopics interface for Seurat objects. For more
detailed guidance on analyzing single-cell RNA-seq data using a topic
model, please see the <a
href="https://stephenslab.github.io/fastTopics/articles">fastTopics
vignettes</a>.</p>
<p>If you find the <strong>fastTopics</strong> package useful for your
work, please cite:</p>
<blockquote>
<p><em>Visualizing the structure of RNA-seq expression data using grade
of membership models.</em></p>
<p>K. K. Dey, C. J. Hsiao and M. Stephens.</p>
<p>PLoS Genetics, 2017.</p>
<p>doi: <a
href="https://doi.org/10.1371/journal.pgen.1006599">10.1371/journal.pgen.1006599</a></p>
</blockquote>
<p>and</p>
<blockquote>
<p><em>Non-negative matrix factorization algorithms greatly improve
topic model fits.</em></p>
<p>P. Carbonetto, A. Sarkar, Z. Wang and M. Stephens.</p>
<p>arXiv, 2021</p>
<p><a href="https://arxiv.org/abs/2105.13440">arXiv 2105.13440</a></p>
</blockquote>
<p>If you use the <strong>de_analysis</strong> function in fastTopics,
please cite:</p>
<blockquote>
<p><em>Interpreting structure in sequence count data with differential
expression analysis allowing for grades of membership.</em></p>
<p>P. Carbonetto, K. Luo, A. Sarkar, A. Hung, K. Tayeb, S. Pott and M.
Stephens</p>
<p>bioRxiv, 2023.</p>
<p>doi: <a
href="https://doi.org/10.1101/2023.03.03.531029">10.1101/2023.03.03.531029</a></p>
</blockquote>
<p>To begin, load the packages we will need to perform the analysis.</p>
<pre class="r"><code>library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(fastTopics)
library(cowplot)</code></pre>
<p>Set the seed so that the results can be reproduced.</p>
<pre class="r"><code>set.seed(1)</code></pre>
<p>Load (and install if necessary) the PBMC 3k data set containing
transcription profiles for 2,700 cells.</p>
<pre class="r"><code>InstallData(&quot;pbmc3k&quot;)
data(pbmc3k)
dim(GetAssayData(pbmc3k))
# [1] 13714  2700</code></pre>
<p>Fit the multinomial topic model to the raw UMI counts. <em>No data
preprocessing or gene filtering is needed.</em> It may take several
minutes to fit the model.</p>
<pre class="r"><code>pbmc3k &lt;- FitTopicModel(pbmc3k, k = 6)</code></pre>
<p>To fit a topic model, we had to choose <span
class="math inline">\(K\)</span>, the number of topics. Here, we chose
<span class="math inline">\(K = 6\)</span> topics. In most settings, a
good choice of <span class="math inline">\(K\)</span> will not be known
in advance, so you will likely want to explore the different settings of
<span class="math inline">\(K\)</span>.</p>
<p>This first plot shows the cells projected onto the top two PCs of the
topic proportions.</p>
<pre class="r"><code>Idents(pbmc3k) &lt;- pbmc3k$seurat_annotations
DimPlot(pbmc3k, reduction = &quot;pca_topics&quot;, pt.size = 1) + theme_cowplot(font_size = 10)</code></pre>
<p><img src="fasttopics_files/figure-html/pca-1-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>Compare this with the top two PCs of the normalized and scaled
counts:</p>
<pre class="r"><code>pbmc3k &lt;- FindVariableFeatures(pbmc3k)
pbmc3k &lt;- NormalizeData(pbmc3k)
pbmc3k &lt;- ScaleData(pbmc3k)
pbmc3k &lt;- RunPCA(pbmc3k)
DimPlot(pbmc3k, reduction = &quot;pca&quot;, pt.size = 1) + theme_cowplot(font_size = 10)</code></pre>
<p><img src="fasttopics_files/figure-html/pca-2-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>The fitted topic model—a “multinom_topic_model” object—is stored in
the “misc” slot:</p>
<pre class="r"><code>fit &lt;- Misc(Reductions(pbmc3k, &quot;multinom_topic_model&quot;))$fit</code></pre>
<p>Once the topic model has been extracted from the Seurat object, many
fastTopics functions can be used for downstream analysis and
visualization. For example, we can create a “Structure plot” to
visualize all <span class="math inline">\(K\)</span> dimensions of the
topic model simultaneously:</p>
<pre class="r"><code>structure_plot(fit, grouping = Idents(pbmc3k), gap = 25)</code></pre>
<p><img src="fasttopics_files/figure-html/structure-plot-1.png" width="900" style="display: block; margin: auto;" /></p>
<p>In the Structure plot, each position along the x-axis is a cell, and
the bar heights at each positioin correspond to topic proportions for
that cell.</p>
<p>From this plot, we observe that topic 3 (green) closely corresponds
to B cells. Topics 1 (red) and 4 (purple) also correspond fairly well to
the two types of monocytes, but there is also some overlap in the
topics; in particular, the purple topic is shared by both types of
monocytes. This suggests that expression in these two types of cells is
less distinguishable than B cells vs. other cell types.</p>
<p>Topic 5 (orange) corresponds closely to natural killer (NK) cells.
Topic 2 (blue) appears to capture biological processes common to T
cells. Interestingly, CD8+ T cells have characteristics of both NK cells
and T cells—these are T cells that sometimes become ``NK-like’’—and this
is reflected in the topic model by assigning membership to both topics 2
and 5 (blue and orange).</p>
<p>The other cell types (platelets, dendritic cells), perhaps because
they are less abundant and/or difficult to distinguish from the other
cell types, are represented as combinations of the topics for the other
cell types.</p>
<p>Topic 6 is present in almost all cells to varying degrees, and
therefore its biological interpretation is not at all clear from the
cell labeling.</p>
<p>TO DO: Illustrate the use of the de_analysis.</p>
<p>This is the version of R and the packages that were used to generate
these results:</p>
<pre class="r"><code>sessionInfo()
# R version 4.2.0 (2022-04-22)
# Platform: x86_64-pc-linux-gnu (64-bit)
# Running under: Red Hat Enterprise Linux 8.4 (Ootpa)
# 
# Matrix products: default
# BLAS/LAPACK: /software/openblas-0.3.13-el8-x86_64/lib/libopenblas_skylakexp-r0.3.13.so
# 
# locale:
#  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
# [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] cowplot_1.1.1           fastTopics_0.6-150      SeuratWrappers_0.3.1   
# [4] pbmc3k.SeuratData_3.1.4 SeuratData_0.2.2        SeuratObject_4.1.3     
# [7] Seurat_4.3.0            rmarkdown_2.14         
# 
# loaded via a namespace (and not attached):
#   [1] plyr_1.8.7             igraph_1.3.1           lazyeval_0.2.2        
#   [4] sp_1.6-0               splines_4.2.0          listenv_0.8.0         
#   [7] scattermore_0.8        ggplot2_3.3.6          digest_0.6.29         
#  [10] invgamma_1.1           htmltools_0.5.2        SQUAREM_2021.1        
#  [13] fansi_1.0.3            magrittr_2.0.3         tensor_1.5            
#  [16] cluster_2.1.3          ROCR_1.0-11            remotes_2.4.2         
#  [19] globals_0.14.0         RcppParallel_5.1.5     matrixStats_0.62.0    
#  [22] R.utils_2.11.0         MCMCpack_1.6-3         spatstat.sparse_3.0-0 
#  [25] prettyunits_1.1.1      colorspace_2.0-3       rappdirs_0.3.3        
#  [28] ggrepel_0.9.1          xfun_0.30              dplyr_1.0.9           
#  [31] crayon_1.5.1           jsonlite_1.8.0         progressr_0.10.0      
#  [34] spatstat.data_3.0-0    survival_3.3-1         zoo_1.8-10            
#  [37] glue_1.6.2             polyclip_1.10-0        gtable_0.3.0          
#  [40] MatrixModels_0.5-0     leiden_0.3.10          future.apply_1.9.0    
#  [43] abind_1.4-5            SparseM_1.81           scales_1.2.0          
#  [46] DBI_1.1.2              spatstat.random_3.1-3  miniUI_0.1.1.1        
#  [49] Rcpp_1.0.9             progress_1.2.2         viridisLite_0.4.0     
#  [52] xtable_1.8-4           reticulate_1.24        rsvd_1.0.5            
#  [55] truncnorm_1.0-8        htmlwidgets_1.5.4      httr_1.4.2            
#  [58] RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-2             
#  [61] farver_2.1.0           pkgconfig_2.0.3        R.methodsS3_1.8.1     
#  [64] sass_0.4.1             uwot_0.1.14            deldir_1.0-6          
#  [67] utf8_1.2.2             labeling_0.4.2         tidyselect_1.1.2      
#  [70] rlang_1.1.0            reshape2_1.4.4         later_1.3.0           
#  [73] munsell_0.5.0          tools_4.2.0            cli_3.3.0             
#  [76] generics_0.1.2         ggridges_0.5.3         evaluate_0.15         
#  [79] stringr_1.4.0          fastmap_1.1.0          yaml_2.3.5            
#  [82] goftest_1.2-3          mcmc_0.9-7             knitr_1.39            
#  [85] fitdistrplus_1.1-8     purrr_0.3.4            RANN_2.6.1            
#  [88] pbapply_1.5-0          future_1.25.0          nlme_3.1-157          
#  [91] mime_0.12              quantreg_5.93          formatR_1.12          
#  [94] R.oo_1.24.0            compiler_4.2.0         plotly_4.10.0         
#  [97] png_0.1-7              spatstat.utils_3.0-1   tibble_3.1.7          
# [100] bslib_0.3.1            stringi_1.7.6          highr_0.9             
# [103] lattice_0.20-45        Matrix_1.5-3           vctrs_0.4.1           
# [106] pillar_1.7.0           lifecycle_1.0.1        BiocManager_1.30.20   
# [109] spatstat.geom_3.0-6    lmtest_0.9-40          jquerylib_0.1.4       
# [112] RcppAnnoy_0.0.19       data.table_1.14.4      irlba_2.3.5           
# [115] httpuv_1.6.5           patchwork_1.1.1        R6_2.5.1              
# [118] promises_1.2.0.1       KernSmooth_2.23-20     gridExtra_2.3         
# [121] parallelly_1.31.1      codetools_0.2-18       MASS_7.3-56           
# [124] assertthat_0.2.1       sctransform_0.3.5      hms_1.1.1             
# [127] parallel_4.2.0         quadprog_1.5-8         grid_4.2.0            
# [130] tidyr_1.2.0            coda_0.19-4            ashr_2.2-54           
# [133] Rtsne_0.16             mixsqp_0.3-48          spatstat.explore_3.0-6
# [136] shiny_1.7.1</code></pre>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
