---
title: "Analyzing Seurat data using fastTopics"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output:
  github_document:
    html_preview: true
    toc: true
    toc_depth: 3
  html_document:
    df_print: kable
---

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120,tidy = TRUE,
                      tidy.opts = list(width.cutoff = 95))
```

This vignette illustrates the use of the [fastTopics][fasttopics]
Seurat wrapper to analyze a Seurat data set. This vignette is only
intended to introduce the basic fastTopics interface for Seurat
objects---for more guidance on analyzing single-cell RNA-seq data
using a topic model, please see the
[fastTopics vignettes][fasttopics-vignettes].

If you find the **fastTopics** package useful for your work, please
cite:

> *Visualizing the structure of RNA-seq expression data using grade of
> membership models.*
>
> K. K. Dey, C. J. Hsiao and M. Stephens.
>
> PLoS Genetics, 2017.
>
> doi: [10.1371/journal.pgen.1006599](https://doi.org/10.1371/journal.pgen.1006599)

and

> *Non-negative matrix factorization algorithms greatly improve topic
> model fits.*
>
> P. Carbonetto, A. Sarkar, Z. Wang and M. Stephens.
>
> arXiv, 2021
>
> [arXiv 2105.13440](https://arxiv.org/abs/2105.13440)

If you use the **de_analysis** function in fastTopics, please cite:

> *Interpreting structure in sequence count data with differential
> expression analysis allowing for grades of membership.*
> 
> P. Carbonetto, K. Luo, A. Sarkar, A. Hung, K. Tayeb, S. Pott and
> M. Stephens
>
> bioRxiv, 2023.
>
> doi: [10.1101/2023.03.03.531029](https://doi.org/10.1101/2023.03.03.531029)

To begin, load the packages we will need to perform the analysis.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(fastTopics)
library(cowplot)
```

Set the seed so that the results can be reproduced.

```{r set-seed}
set.seed(1)
```

Load (and install if necessary) the PBMC 3k data set containing
transcription profiles for 2,700 cells.

```{r load-data, warning=FALSE}
InstallData("pbmc3k")
data(pbmc3k)
dim(GetAssayData(pbmc3k))
```

Fit the multinomial topic model to the raw UMI counts. *No data
preprocessing or gene filtering is needed.* It may take several
minutes to fit the model.

```{r fit-model, results="hide", message=FALSE}
pbmc3k <- FitTopicModel(pbmc3k,k = 6)
```

To fit a topic model, we had to choose $K$, the number of topics.
Here, we chose $K = 6$ topics. In most settings, a good choice of $K$
will not be known in advance, so you will likely want to explore the
different settings of $K$.

This first plot shows the cells projected onto the top two PCs of the
topic proportions.

```{r pca-1, fig.width=4.5, fig.height=3}
Idents(pbmc3k) <- pbmc3k$seurat_annotations
DimPlot(pbmc3k,reduction = "pca_topics",pt.size = 1) +
  theme_cowplot(font_size = 10)
```

Compare this with the top two PCs of the normalized and scaled counts:

```{r pca-2, fig.width=4.5, fig.height=3, message=FALSE}
pbmc3k <- FindVariableFeatures(pbmc3k)
pbmc3k <- NormalizeData(pbmc3k)
pbmc3k <- ScaleData(pbmc3k)
pbmc3k <- RunPCA(pbmc3k)
DimPlot(pbmc3k,reduction = "pca",pt.size = 1) +
  theme_cowplot(font_size = 10)
```

The fitted topic model---a "multinom_topic_model" object---is stored
in the "misc" slot:

```{r extract-fit}
fit <- Misc(Reductions(pbmc3k,"multinom_topic_model"))
```

Once the topic model has been extracted from the Seurat object, many
fastTopics functions can be used for downstream analysis and
visualization. For example, we can create a *Structure plot* to
visualize all $K$ dimensions of the topic model simultaneously:

```{r structure-plot, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
structure_plot(fit,grouping = Idents(pbmc3k),gap = 25)
```

In this plot, each position along the x-axis is a cell, and bar
heights correspond to topic proportions (which sum to 1 for each cell).

TO DO: Illustrate the use of the de_analysis.

This is the version of R and the packages that were used to generate
these results:

```{r session-info}
sessionInfo()
```

[fasttopics]:  https://github.com/stephenslab/fastTopics
[fasttopics-vignettes]:  https://stephenslab.github.io/fastTopics/articles
