---
title: "Analyzing Seurat data using fastTopics"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output:
  github_document:
    html_preview: true
  html_document:
    df_print: kable
    highlight: textmate
    theme: readable
---

Here we illustrate the use of the [fastTopics][fasttopics] Seurat
wrapper to analyze a Seurat data set. This vignette is only intended
to introduce the basic fastTopics interface for Seurat objects---for
background and guidance on analysis of single-cell RNA-seq data using
a topic model, please see the
[fastTopics vignettes][fasttopics-vignettes].

If you find the **fastTopics** package useful for your work, please
cite:

Kushal K. Dey, Chiaowen Joyce Hsiao and Matthew Stephens (2017).
[Visualizing the structure of RNA-seq expression data using grade of
membership models.][countclust-paper] *PLoS Genetics* **13**, e1006599.

Peter Carbonetto, Kevin Luo, Kushal Dey, Joyce Hsiao and Matthew
Stephens (2021). fastTopics: fast algorithms for fitting topic models
and non-negative matrix factorizations to count data. R package
version 0.4-23. [https://github.com/stephenslab/fastTopics][fasttopics]

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

We begin by loading the packages used to perform the analysis.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(fastTopics)
library(cowplot)
```

We set the seed so that the results can be reproduced.

```{r set-seed}
set.seed(1)
```

Load---and, if necessary, install---the PBMC 3k data set containing
transcription profiles for 2,700 cells.

```{r load-data, warning=FALSE}
InstallData("pbmc3k")
data(pbmc3k)
dim(GetAssayData(pbmc3k))
```

Fit the multinomial topic model to the raw UMI counts---*no
pre-processing or pre-selection of genes is needed.* Note that it may
take several minutes to complete this model fitting step.

```{r fit-model, results="hide", message=FALSE, cache=TRUE}
pbmc3k <- FitTopicModel(pbmc3k,k = 6)
```

To fit a topic model, we must specify $K$, the number of topics.
Here, we have chosen $K = 6$ topics. In most settings, a good choice
of $K$ will not be known in advance, so you will you want to explore
the results from topic models at different settings of $K$.

This plot shows the cells projected onto the top two principal
components (PCs) of the topic mixture proportions.

```{r pca-1, fig.width=4.5, fig.height=3}
Idents(pbmc3k) <- pbmc3k$seurat_annotations
DimPlot(pbmc3k,reduction = "pca_topics",pt.size = 1) +
  theme_cowplot(font_size = 10)
```

Compare this against the top two PCs of the transformed counts:

```{r pca-2, fig.width=4.5, fig.height=3, message=FALSE}
pbmc3k <- FindVariableFeatures(pbmc3k)
pbmc3k <- ScaleData(pbmc3k)
pbmc3k <- RunPCA(pbmc3k)
DimPlot(pbmc3k,reduction = "pca",pt.size = 1) +
  theme_cowplot(font_size = 10)
```

The fitted topic model---a "multinom_topic_model" object---is stored
in the "misc" slot:

```{r extract-fit}
fit <- Misc(Reductions(pbmc3k,"multinom_topic_model"))
```

Once the fitted topic model is extracted, many functions from the
**fastTopics** package can be used for analysis and visualization. For
example, the Structure plot provides an evocative visual summary of
the estimated mixture proportions for each cell. Here, we have grouped
the cells by previously inferred labels.

```{r structure-plot, fig.width=7.5, fig.height=1.75, results="hide", message=FALSE}
structure_plot(fit,grouping = Idents(pbmc3k),gap = 25)
```

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```

[fasttopics]:  https://github.com/stephenslab/fastTopics
[fasttopics-vignettes]:  https://stephenslab.github.io/fastTopics/articles
[countclust-paper]: https://doi.org/10.1371/journal.pgen.1006599
