---
title: "Running scGSEA from gficf on Seurat Objects"
date: 'Compiled: `r format(Sys.Date(), "%m %d %y")`'
output: 
  html_document:
    df_print: kable
  github_document:
    html_preview: true
    toc: true
    toc_depth: 3
---

This vignette demonstrates how to run single-cell gene set enrichment analysis with gficf on Seurat objects. 

> *Single-cell gene set enrichment analysis and transfer learning for functional annotation of scRNA-seq data*
>
> Franchini Melania, Pellecchia Simona , Viscido Gaetano, Gambardella Gennaro
> 
> NAR Genomics and Bioinformatics, Volume 5, Issue 1, March 2023, lqad024
> 
> doi:
> <https://doi.org/10.1093/nargab/lqad024>
> 
> GitHub:
> <https://github.com/gambalab/gficf>

```{r setup, include=FALSE}

knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.height = 10,
  fig.width = 20
)

```


Prerequisites to install: 

* [Seurat](https://satijalab.org/seurat/install) 
* [SeuratData](https://github.com/satijalab/seurat-data) 
* [SeuratWrappers](https://github.com/satijalab/seurat-wrappers)
* [gficf](https://github.com/gambalab/gficf)

```{r looad libraries}

require(Seurat)
require(SeuratData)
require(SeuratWrappers)
require(gficf)

```

# Installation
Please install the Seurat wrapper which enables to run single-cell gene set enrichment analysis with gficf on Seurat objects as reported here:

```{r install wrapper from gficf}

if(!require(devtools)){install.packages("devtools")}
devtools::install_github("gambalab/seurat-wrappers")

```


# Introduction
GF-ICF is an R package for normalization, visualization and analysis of of single-cell RNA sequencing data, based on a data transformation model called term frequency–inverse document frequency (TF-IDF), which has been extensively used in the field of text mining. Here, we show one of the latest implementation which enables the pathway activity computation at single-cell level.

## Load examle data from SeuratData
To demonstrate how to run runScGSEA() function on a single-cell RNA-seq dataset.
We will use the pbmc3k.final dataset from the SeuratData package which contains 2,638 PBMCs processed using the standard Seurat workflow as demonstrated in Seurat's guided clustering tutorial.

```{r load example data}

InstallData("pbmc3k")
data("pbmc3k.final")
pbmc3k.final

```

## Compute pathways activity at single-cell resolution
Single cell gene set enrichment analysis is performed by the function runScGSEA() of gficf package which computes GSEA scores for each cell across a set of input pathways by using NMF.
The list of gene sets to use can can be specified trough the category and subcategory parameters. 

```{r runScGsea, results='hide'}

pbmc.ptw <- RunScGSEA(object = pbmc3k.final, geneID = "symbol",  species = "human", category = "H")

```



# PART 1 : Evaluate pathways activity in single-cell data
## Visualize UMAP plot computed on genes expression
As dimensionality reduction results computed on genes expression are preserved, let's graph the UMAP computed on genes x cells matrix. In this case cells are colored by their cell type.

```{r DimPlotGenes}

DimPlot(pbmc.ptw, reduction = "umap_byGenes", group.by = "seurat_annotations")

```

## Identify how pathways activity change across cell type
To evaluate how pathways activity change across cell type, we utilize DotPlot() function where dot size encodes for the percentage of cells of each cell type in which the given pathway is active. Finally, cell types are ordered (by hierarchical clustering) based on given features by setting cluster.idents = TRUE.

```{r DotPlotPathways}

library(ggplot2)
library(RColorBrewer)

all.pathways = rownames(pbmc.ptw)

DotPlot(object = pbmc.ptw,
        features = all.pathways,
        split.by = 'seurat_annotations',
        cols = RColorBrewer::brewer.pal(n = length(unique(pbmc.ptw$seurat_annotations)), name = 'Set1'),
        cluster.idents = TRUE) +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust = 1, size = 10))

```

## Inspect pathways activation in each cell of a given cell type
To inspect pathways activation in each cell of a given cell type, we use DoHeatmap() function. In this case, we are plotting all given pathways for each cell type.

```{r DoHeatmapPathways}

DoHeatmap(pbmc.ptw, features = all.pathways, group.by = "seurat_annotations", angle = 90) + NoLegend()

```



# PART 2 : Analyze data based on pathways activity scores
## Visualize UMAP plot computed on pathways activity scores
Now, let's see the UMAP computed on pathways activity scores. To this end, we follow the standard Seurat workflow which requires the computation of PCA on scaled data (i.e. pathway scores normalized by gene-set) before run the UMAP reduction. Results will be stored in separate reduction slot of the Seurat object.
Finally, we display the UMAP plot where cells are colored by their cell type.

```{r UMAPcellTypePathways, results='hide'}

pbmc.ptw <- RunPCA(pbmc.ptw, features = all.pathways)
pbmc.ptw <- RunUMAP(pbmc.ptw, dims = 1:5)
DimPlot(pbmc.ptw, reduction = "umap", group.by = 'seurat_annotations')

```

## Cluster cells based on pathways activity scores
Now that we reconstructed pathway’s activity at single cell level we can try to cluster cell according to these values using Seurat functions FindNeighbors() and FindClusters(). Finally, we graph clustering results computed both on gene expression and pathway activity scores as UMAP plot form.
```{r UMAPclustersPathways}

pbmc.ptw <- FindNeighbors(pbmc.ptw, dims = 1:10)
pbmc.ptw <- FindClusters(pbmc.ptw, resolution = 0.5)
DimPlot(pbmc.ptw, reduction = "umap", group.by = c('seurat_clusters_byGenes', 'seurat_clusters'))

```


# TIPS
## How can I handle collection with an high number of genesets?
In the case you are handling collections containing a lot of gene-sets, it could be useful to perform a feature selection step. To this end, we first store pathways metadata in a separate variable (i.e. feat.meta), then we create an empty dataframe in the meta.features slot to enable the storing of the output of the Seurat function FindVariableFeatures() setting nfeatures equal to 10% of the number of genesets used.

```{r PCAselection, results='hide'}

pbmc.ptw.kegg <- RunScGSEA(object = pbmc3k.final,
                           geneID = "symbol",
                           species = "human",
                           category = "C2",
                           subcategory = 'CP:KEGG')

feat.meta <- pbmc.ptw.kegg[['RNA']]@meta.features
pbmc.ptw.kegg[['RNA']]@meta.features <- data.frame(row.names = rownames(pbmc.ptw.kegg))
nfeatures <- round(nrow(pbmc.ptw.kegg)*0.10)
pbmc.ptw.kegg <- FindVariableFeatures(pbmc.ptw.kegg, selection.method = "vst", nfeatures = nfeatures)
top <- head(VariableFeatures(pbmc.ptw.kegg), nfeatures)
LabelPoints(plot = VariableFeaturePlot(pbmc.ptw.kegg), points = top, repel = TRUE)

pbmc.ptw.kegg <- RunPCA(pbmc.ptw.kegg, features = VariableFeatures(object = pbmc.ptw.kegg))
VizDimLoadings(pbmc.ptw.kegg, dims = 1:2, reduction = "pca")

```

